swarm/src/handler.rs

```diff
@@ -459,54 +459,67 @@ impl<TConnectionUpgrade, TOutboundOpenInfo, TCustom, TErr>
     }
 }
 
+#[deprecated(note = "Renamed to `StreamUpgradeError`")]
+pub type ConnectionHandlerUpgrErr<TUpgrErr> = StreamUpgradeError<TUpgrErr>;
+
 /// Error that can happen on an outbound substream opening attempt.
 #[derive(Debug)]
-pub enum ConnectionHandlerUpgrErr<TUpgrErr> {
+pub enum StreamUpgradeError<TUpgrErr> {
     /// The opening attempt timed out before the negotiation was fully completed.
     Timeout,
-    /// Error while upgrading the substream to the protocol we want.
-    Upgrade(UpgradeError<TUpgrErr>),
+    /// The upgrade produced an error.
+    Apply(TUpgrErr),
+    /// No protocol could be agreed upon.
+    NegotiationFailed,
```

Comment:
As a side-note, we had multiple users complain, that it is hard to debug why the negotiation failed. What would have helped them is `NegotiationFailed` to contain the protocols the local and the remote peer offered.

That said, using `DEBUG` logging and looking for the `multistream-select` lines also helps the debugging, though a bit harder to discover.

Not to be addressed in this pull request.

Reply:
Including this information will be a breaking change but we can introduce the facilities for it in a non-breaking manner once we have completely encapsulated `multistream-select`, meaning this is a step in the right direction I tihnk.